transform_aes.c

```diff
@@ -196,62 +236,174 @@ static int transop_encode_aes( n2n_trans_op_t * arg,
+	    break;
+	}
+	case 2 :
+	{
+	    ethernet_data_offset = TRANSOP_AES_DATA_LEN;
+	    /* version 2 also requires the data length to be part of encrypted data */
+	    data_len = (uint16_t *)assembly;
+	     *data_len = htons(in_len);
+	    buffer_suffix = 0; /* nothing to be added behind buffer's end */
+	    skip_to_data = TRANSOP_AES_VER_SIZE + TRANSOP_AES_SA_SIZE + N2N_AES_MSGAUTH_SIZE + N2N_AES_IVEC_SIZE;
+	    break;
+	}
+    }
+    /* also count the prepending data fields for encryption */
+    len = in_len + ethernet_data_offset;
+    /* Round up to next whole AES block size */
+    len2 = ( ( ( len - 1 + buffer_suffix) / AES_BLOCK_SIZE) + 1 ) * AES_BLOCK_SIZE; 
+
+    /* The assembly buffer is a source for encrypting data.
+     * The whole contents of assembly are encrypted. */
+    memcpy( assembly + ethernet_data_offset , inbuf, in_len );
+
+    /* prepare the encryption by performing the version-dependant padding
+       at buffer's end as well as the IV, also choose key */
+    AES_KEY key;
+    switch (priv->version)
+    {
+	case 1 :
+ 	{
+	    assembly[ len2-1 ]=(len2-len);
             traceEvent( TRACE_DEBUG, "padding = %u", assembly[ len2-1 ] );
+	    /* set IV to 0 for v1 */
+	    memset (iv, 0, N2N_AES_IVEC_SIZE);
+            /* data will be encrypted using v1 key K */
+	    key = sa->enc_key_K;
+	    break;
+	}
+	case 2 :
+	{
+	    /* generate random padding for the last plain text block and also
+               generate a random initialization vector for CBC directly to its field in the buffer
+               only proceed if random values are cryptographically safe (return value == 1) */
+	    if ((RAND_bytes(assembly + len + 1, len2 - len) != 1) ||
+	        (RAND_bytes(outbuf + TRANSOP_AES_VER_SIZE + TRANSOP_AES_SA_SIZE + N2N_AES_MSGAUTH_SIZE, N2N_AES_IVEC_SIZE) != 1))
+	    {
+	        traceEvent ( TRACE_ERROR, "encode_aes no random data available for padding and initialization vector." );
+	        goto error;
+	    }
+	    /* encode system time to IV for replay attack protection */
+	    /* get current system time (for international use: not local time) to encode it in IV */
+	    struct timeval tod;
+	    uint64_t micro_seconds;
+	    gettimeofday (&tod, NULL);
+	    /* We will calculate and put the microseconds since 1970 leftbound into the IV.
+	       As microseconds fraction never exceeds 1,000,000 , a max of 20 bits is
+	       used to encode the value tv_usec. 32 bits are used for tv_sec.
+	       Thus, micro_seconds' 12 most significant bits are not used.
+	       As we do not want 12 definetely zeroed bits IV, some bit masking is required
+	       to only copy the 52 (least) significant bits leaving the other 12 random as before
+	       - converting in host byte order to make sure bitwise masking is applied correctly;
+	       shifting might be independet from byte order, though. */
+	    micro_seconds = be64toh(*(uint64_t*)(outbuf + TRANSOP_AES_VER_SIZE + TRANSOP_AES_SA_SIZE + N2N_AES_MSGAUTH_SIZE));
```

Comment:
Please also consider this issue with 32bit seconds: https://en.wikipedia.org/wiki/Year_2038_problem . Since we have extra bits it's a waste not to use them and instead make they random. Moreover, do we really need microseconds accuracy here? What if the clocks of the two machines are not synchronized?

Reply:
I actually have thought about it. What will happen is that the time counter turns over to zero. A reboot/restart after that will fix it.

IVs should not just not repeat under the same key, they also should be as unpredictable as possible. That is achieved by 2 means: Using as much randomness as space allows (thus only using the neccessary amount of the highly predictable time bits) and the following encryption step using K1 before feeding the IV to the cbc encryption.

The time stamp part serves two purposes in replay attack protection:

1. It makes sure that the packet was sent not too long ago. Taking into account that packets may take some time and also, that clocks might be not completely in sync, there is a grace time of +/- 16 seconds (= 32 seconds) for comparison of clocks of _different computers_, specified at the top of transform_aes. It is user's deliberately made decision to use V2 by setting the parameter. The user should be aware that higher security requirements such as replay attack protection also have some boundary conditions such as somewhat synchronous clocks in this case, I will add a section to the man page. In times of ubiquitous ntp servers, I think this is a feasible approach. Let me know what you think.

2. By comparing the packet time stamps to each other, i. e. comparing time stamps from the _same clock_, we make sure that packets are in sequence. The microseconds are a kind of counter. To be specific, we make sure that they are "somewhat" in sequence as we allow for some grace time of 160 ms because packets sometimes may overtake each other on the line or even the packet stack. This time is defined at the top of transform_aes, too. In my worldwide setup, this value yielded good results. The purpose is to prevent fast replay attacks inside the 32 seconds window mentioned above. That's what we need the high accuracy for. If tests show that 160 milliseconds are a sufficent reasonable grace time, we could lower the accuracy of the time stamp further by the factor of 128...1024  microseconds by shifting it 7...10 bits further to the right and allowing more space for random bits at the same time.